draft-ietf-mls-protocol.md

```diff
@@ -1435,44 +1465,230 @@ all arrive at the following state:
+with Nonces and keys generated from the same Application secrets.
+
+[[OPEN ISSUE: The AEAD key exhaustion bound applies on a per Application
+Secret basis. This depends on the number of messages per Application secrets
+that we allow. Should we set a limit ?
+]]
+
+Note that each change to the Group through a Handshake message will cause
+a change of the Group Secret. Hence this change MUST be applied before encrypting
+any new Application message. This is required for confidentiality reasons
+in order for Members to avoid receiving messages from the group after leaving or
+being added to, or excluded from the Group.
+
+## Application Key Schedule {#key-schedule-application}
+
+Updating the Application secret and deriving the associated AEAD key and nonce can
+be summarized as the following Application key schedule where
+each participant's Application secret chain looks as follows after the initial
+derivation:
 
-The primary purpose of this protocol is to enable an authenticated
-group key exchange among participants.  In order to protect messages sent among
-those participants, an application will need to specify how messages are protected.
+~~~~~
+           application_secret_N-1
+                     |
+                     +--> HKDF-Expand-Label(.,"nonce", "", nonce_length)
+                     |    = write_nonce_N-1_[sender]
+                     |
+                     +--> HKDF-Expand-Label(.,"key", "", key_length)
+                     |    = write_key_N-1_[sender]
+                     |
+                     V
+           Derive-Secret(., "app upd","")
+                     |
+                     V
+           application_secret_N
+                     |
+                     +--> HKDF-Expand-Label(.,"nonce", "", nonce_length)
+                     |    = write_nonce_N_[sender]
+                     |
+                     +--> HKDF-Expand-Label(.,"key", "", key_length)
+                          = write_key_N_[sender]
+
+~~~~~
+
+The Application context provided together with the previous Application secret
+is used to bind the Application messages with the next key and add some freshness.
+
+[[OPEN ISSUE: Context is left empty for now, it might be that using only
+the message counter is enough, this would be more conveniant.
+Hashing all the data is obviously very costly and prevents from encrypt in
+parallel, an other solution could be to add a pseudo-random contribution to
+each message and only hash these. ]]
+
+[[ OPEN ISSUE: At the moment there is no contributivity of Application secrets
+chained from the initial one to the next generation of Epoch secret. While this
+seems safe because cryptographic operations using the application secrets can't
+affect the group init_secret, it remains to be proven correct. ]]
+
+### Updating the Application Secret
+
+When the Application Secret is updated:
+
+- Senders MUST use the generation N+1 of the application secret, where N is
+the last generation they received.
```

Comment:
There is no good reason for this restriction. It doesn't make the key schedule any easier.